packages/medusa/src/services/event-bus.ts

```diff
@@ -241,30 +275,84 @@ export default class EventBusService {
    * @param job The job object
    * @return resolves to the results of the subscriber calls.
    */
-  worker_ = async <T>(job: {
-    data: { eventName: string; data: T }
-  }): Promise<unknown[]> => {
+  worker_ = async <T>(job: BullJob<T>): Promise<unknown> => {
     const { eventName, data } = job.data
-    const eventObservers = this.observers_.get(eventName) || []
-    const wildcardObservers = this.observers_.get("*") || []
+    const eventSubscribers = this.eventToSubscribersMap_.get(eventName) || []
+    const wildcardSubscribers = this.eventToSubscribersMap_.get("*") || []
+
+    const allSubscribers = eventSubscribers.concat(wildcardSubscribers)
 
-    const observers = eventObservers.concat(wildcardObservers)
+    // Pull already completed subscribers from the job data
+    const completedSubscribers = job.data.completedSubscriberIds || []
 
-    this.logger_.info(
-      `Processing ${eventName} which has ${eventObservers.length} subscribers`
+    // Filter out already completed subscribers from the all subscribers
+    const subscribersInCurrentAttempt = allSubscribers.filter(
+      (observer) => observer.id && !completedSubscribers.includes(observer.id)
     )
 
-    return await Promise.all(
-      observers.map(async (subscriber) => {
-        return subscriber(data, eventName).catch((err) => {
-          this.logger_.warn(
-            `An error occurred while processing ${eventName}: ${err}`
-          )
-          console.error(err)
-          return err
-        })
+    const isRetry = job.attemptsMade > 0
+
+    if (isRetry) {
+      this.logger_.info(
+        `Retrying (attempt ${job.attemptsMade}) ${eventName} which has ${eventSubscribers.length} subscribers (${subscribersInCurrentAttempt.length} of them failed)`
+      )
+    } else {
+      this.logger_.info(
+        `Processing ${eventName} which has ${eventSubscribers.length} subscribers`
+      )
+    }
+
+    const completedSubscribersInCurrentAttempt: string[] = []
+
+    const subscribersResult = await Promise.all(
```

Comment:
self-review: @adrien2p I decided to store the result of the successful promises to be able to return them in the end. That way, the API of the worker is kept as is.

Reply:
Yeah that works as well. I like it too 👌